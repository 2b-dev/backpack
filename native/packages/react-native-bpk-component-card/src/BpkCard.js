/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2017 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  View,
  Platform,
  StyleSheet,
  TouchableHighlight,
} from 'react-native';
import React from 'react';
import PropTypes from 'prop-types';

const tokens = Platform.select({
  ios: () => require('bpk-tokens/tokens/ios/base.react.native.common.js'), // eslint-disable-line global-require
  android: () => require('bpk-tokens/tokens/android/base.react.native.common.js'), // eslint-disable-line global-require
})();

const {
  colorGray50,
  colorGray900,
  colorWhite,
  spacingBase,
} = tokens;

/**
 * Define styles needed by the component
 */
const styles = StyleSheet.create({
  card: {
    // TODO: update once tokens are available
    backgroundColor: colorWhite,
    elevation: 3,
    shadowColor: colorGray900,
    shadowOffset: { width: 0, height: 1 },
    shadowOpacity: 0.3,
    shadowRadius: 3,
  },
  common: {
    // TODO: update once tokens are available
    borderRadius: 8,
  },
  overflowHidden: {
    // TODO: update once tokens are available
    overflow: 'hidden',
  },
  padded: {
    padding: spacingBase,
  },
});

const BpkCard = (props) => {
  const { padded, children, onPress, style, ...rest } = props;

  const cardBaseStyle = [styles.card, styles.common];

  if (padded) {
    cardBaseStyle.push(styles.padded);
  }

  return (
    <TouchableHighlight
      accessibilityComponentType="button"
      onPress={onPress}
      style={[cardBaseStyle, style]}
      underlayColor={colorGray50}
      {...rest}
    >
      <View
        style={[styles.overflowHidden, styles.common]}
      >
        {children}
      </View>
    </TouchableHighlight>
  );
};

BpkCard.propTypes = {
  children: PropTypes.node.isRequired,
  onPress: PropTypes.func.isRequired,
  padded: PropTypes.bool,
  style: View.propTypes.style,
};

BpkCard.defaultProps = {
  children: null,
  padded: true,
  style: null,
};

export default BpkCard;
