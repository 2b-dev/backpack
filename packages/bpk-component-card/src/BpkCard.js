import React, { PropTypes } from 'react'

import './bpk-card.scss'

const BpkCard = props => {
  const classNames = [ 'bpk-card' ]

  props.padded ? classNames.push('bpk-card--padded') : null

  return <a href={props.href} onClick={props.onClick} className={classNames.join(' ')}>{props.children}</a>
}

BpkCard.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
  href: PropTypes.string.isRequired,
  onClick: PropTypes.func,
  padded: PropTypes.bool
}

BpkCard.defaultProps = {
  onClick: null,
  padded: true
}

export default BpkCard
