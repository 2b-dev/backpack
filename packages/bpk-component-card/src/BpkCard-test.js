import React from 'react'
import renderer from 'react/lib/ReactTestRenderer'
import BpkCard from './BpkCard'

describe('BpkCard', () => {
  it('should render correctly', () => {
    const tree = renderer.create(
      <BpkCard href='//www.skyscanner.net'>
        Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum
        sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
      </BpkCard>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render correctly with "tagName" attribute equal to "article"', () => {
    const tree = renderer.create(
      <BpkCard href='//www.skyscanner.net' tagName='article'>
        Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum
        sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
      </BpkCard>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render correctly with "padded" attribute equal to "false"', () => {
    const tree = renderer.create(
      <BpkCard href='//www.skyscanner.net' padded={false}>
        Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum
        sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
      </BpkCard>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render correctly with "focused" attribute', () => {
    const tree = renderer.create(
      <BpkCard href='//www.skyscanner.net' focused>
        Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum
        sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
      </BpkCard>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})

