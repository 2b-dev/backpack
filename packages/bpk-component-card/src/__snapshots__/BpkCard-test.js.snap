exports[`BpkCard should render correctly 1`] = `
<a
  className="bpk-card bpk-card--padded"
  href="//www.skyscanner.net"
  onClick={null}>
  Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
</a>
`;

exports[`BpkCard should render correctly with "focused" attribute 1`] = `
<a
  className="bpk-card bpk-card--padded bpk-card--focused"
  href="//www.skyscanner.net"
  onClick={null}>
  Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
</a>
`;

exports[`BpkCard should render correctly with "padded" attribute equal to "false" 1`] = `
<a
  className="bpk-card"
  href="//www.skyscanner.net"
  onClick={null}>
  Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
</a>
`;

exports[`BpkCard should render correctly with "tagName" attribute equal to "article" 1`] = `
<a
  className="bpk-card bpk-card--padded"
  href="//www.skyscanner.net"
  onClick={null}>
  Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
</a>
`;
